template/runtime/src/lib.rs

```diff
@@ -680,6 +683,9 @@ impl_runtime_apis! {
 				access_list.unwrap_or_default(),
 				is_transactional,
 				validate,
+				// TODO we probably want to support external cost recording in non-transactional calls
```

Comment:
It's not a matter of validating in the non-transactional context, but appropriately `OutOfGas` if the pov size is exhausted. Otherwise gas estimation will break.